### PR TITLE
[Feature] concurrently jit and put compiled func into LRU cache

### PR DESCRIPTION
**Pull request by fzhedu**
Why I'm doing:

cache the compiled func to avoid compile again for identical expressions, as well as manage memory of jit.

What I'm doing:

1. put compiled func into LRU cache, managed by used memory, which is the size of memory buffer of each LLVM module.
2. imp jit_func_name() to get short names, reducing memory useage, instead of using debug_string().
3. support concurrently jit by creating a jit engine for compiling each expression.
4. reformat LLVM including files.
5. LRU cache can fix following issue.

Fixes https://github.com/StarRocks/StarRocksTest/issues/6101

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/jit/jit_engine.cpp [hunk 1]
```diff
@@ -74,6 +85,11 @@ Status JITEngine::init() {
 
     _initialized = true;
     _support_jit = true;
+#if BE_TEST
+    _func_cache = new_lru_cache(32); // 1 capacity per cache of 32
+#else
+    _func_cache = new_lru_cache(3200); // 100 capacity per cache of 32
```
Comment (imay) [inline be/src/exprs/jit/jit_engine.cpp:91]: Better to be configurable.

Comment (imay) [reply be/src/exprs/jit/jit_engine.cpp:91]: And what does it mean by 100 capacity per cache?

File: be/src/common/config.h [hunk 2]
```diff
@@ -1195,4 +1195,7 @@ CONF_mInt64(pk_dump_interval_seconds, "3600"); // 1 hour
 
 // whether enable query profile for queries initiated by spark or flink
 CONF_mBool(enable_profile_for_external_plan, "false");
+
+// jit LRU cache size, will div 32 shards, i.e., 100 per shard by default, taking more memory if larger
```
Comment (imay) [inline be/src/common/config.h:1199]: Could you comment on how much memory it takes when it is set to default?

Comment (fzhedu) [reply be/src/common/config.h:1199]: it depends on complex expressions, we will test memory usage later. As commented, we will evit from cache by the memory usage in the next work, I don't find a better way to collect memory usage now.

Comment (fzhedu) [reply be/src/common/config.h:1202]: yes, I will fix the memory issue soon after passing the CI.

**Comment by andyziye** [comment]
@Mergify rebase

**Comment by andyziye** [comment]
@mergify rebase

File: be/src/exprs/arithmetic_expr.cpp [hunk 3]
```diff
@@ -189,6 +195,11 @@ class VectorizedDivArithmeticExpr final : public Expr {
 
     bool is_compilable() const override { return Type != TYPE_LARGEINT && IRHelper::support_jit(Type); }
 
+    std::string jit_func_name() const override {
```
Comment (satanson) [inline be/src/exprs/arithmetic_expr.cpp:198]: Have you conducted some profile test on jit_func_name(), it seems inefficient, especially for  short-term query that only return a few rows.  can we construct unique function name in a efficient way?

File: be/src/common/config.h [hunk 4]
```diff
@@ -1215,4 +1215,7 @@ CONF_mBool(enable_profile_for_external_plan, "false");
 
 // the max length supported for varchar type
 CONF_mInt32(olap_string_max_length, "1048576");
+
+// jit LRU cache size for total 32 shards.
+CONF_mInt64(jit_lru_cache_size, "1073741824");
```
Comment (silverbullet233) [inline be/src/common/config.h:1220]: 1G may be too large if BE deployed on a small instance, should we support set a ratio for it? just like storage_page_cache_limit

File: be/src/exprs/arithmetic_expr.cpp (hunk 3 above)
Comment (fzhedu) [reply be/src/exprs/arithmetic_expr.cpp:198]: we see it takes tens of us when hiting the cache. It is not a bottleneck now. For short_term query, we can close the jit according to plans cardinality later.

File: be/src/common/config.h (hunk 4 above)
Comment (fzhedu) [reply be/src/common/config.h:1220]: if the cache size is too smaller, the cache becomes invalid, possiblely introducing some bad cases. We can disable jit if the system memory or process  mem limit is less than X bytes.

**Review by satanson — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

File: be/CMakeLists.txt [hunk 5]
```diff
@@ -570,7 +575,11 @@ set (LLVM_LIBRARIES
     LLVMTransformUtils
     LLVMSelectionDAG
     LLVMMCParser
+    LLVMX86AsmParser
```
Comment (kevincai) [inline be/CMakeLists.txt:578]: TODO: support arm build, this will break arm build.

**Review by kevincai — APPROVED** [review]

**Pull request merged**